packages/compiler-cli/src/ngtsc/typecheck/diagnostics/src/diagnostic.ts

```diff
@@ -33,21 +33,26 @@ export interface TemplateDiagnostic extends ts.Diagnostic {
 export function makeTemplateDiagnostic(
     templateId: TemplateId, mapping: TemplateSourceMapping, span: ParseSourceSpan,
     category: ts.DiagnosticCategory, code: number, messageText: string|ts.DiagnosticMessageChain,
-    relatedMessage?: {
+    relatedMessages?: {
       text: string,
-      span: ParseSourceSpan,
-    }): TemplateDiagnostic {
+      start: number,
+      end: number,
+      sourceFile: ts.SourceFile,
+    }[]): TemplateDiagnostic {
   if (mapping.type === 'direct') {
     let relatedInformation: ts.DiagnosticRelatedInformation[]|undefined = undefined;
-    if (relatedMessage !== undefined) {
-      relatedInformation = [{
-        category: ts.DiagnosticCategory.Message,
-        code: 0,
-        file: mapping.node.getSourceFile(),
-        start: relatedMessage.span.start.offset,
-        length: relatedMessage.span.end.offset - relatedMessage.span.start.offset,
-        messageText: relatedMessage.text,
-      }];
+    if (relatedMessages !== undefined) {
+      relatedInformation = [];
+      relatedMessages.forEach(relatedMessage => {
```

Comment:
```suggestion
      for (const relatedMessage of relatedMessages) {
```

Besides being easier to debug, this will also allow TypeScript to better understand how the types are narrowed here, so you shouldn't need to use `?.push()`.

Reply:
You could also consider using `.reduce` here since you're transforming the data while iterating through it. That would look like
```
let relatedInformation = relatedMessages.reduce((infoArr, relatedMessage) => {
       infoArr.push({
          category: ts.DiagnosticCategory.Message,
          code: 0,
          file: relatedMessage.sourceFile,
          start: relatedMessage.start,
          length: relatedMessage.end - relatedMessage.start,
          messageText: relatedMessage.text,
       });
       return infoArr;
}, []);
```